src/sentry/utils/snowflake.py

```diff
@@ -0,0 +1,106 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import OrderedDict, Tuple
+
+from django.conf import settings
+from django.db import IntegrityError
+
+from sentry.utils import redis
+
+_TTL = timedelta(minutes=5)
+SENTRY_EPOCH_START = datetime(2022, 4, 26, 0, 0).timestamp()
+
+
+class SnowflakeIdMixin:
+    def save_with_snowflake_id(self, snowflake_redis_key, save_callback):
+        for _ in range(settings.MAX_REDIS_SNOWFLAKE_RETY_COUNTER):
+            if not self.id:
+                self.id = snowflake_id_generation(snowflake_redis_key)
+            try:
+                save_callback()
+                return
+            except IntegrityError:
+                self.id = None
+        raise Exception("Max allowed ID retry reached. Please try again in a second")
+
+
+@dataclass(frozen=True, eq=True)
+class SnowflakeBitSegment:
+    length: int
+    name: str
+
+    def validate(self, value):
+        if self.length <= 0:
+            raise Exception("The length should be a positive number")
+        if value >> self.length != 0:
+            raise Exception(f"{self.name} exceed max bit value of {self.length}")
+        return True
+
+
+ID_VALIDATOR = SnowflakeBitSegment(settings.SNOWFLAKE_ID_LENGTH, "Snowflake ID")
+
+VERSION_ID = SnowflakeBitSegment(settings.SNOWFLAKE_VERSION_ID_LENGTH, "Version ID")
+TIME_DIFFERENCE = SnowflakeBitSegment(settings.SNOWFLAKE_TIME_DIFFERENCE_LENGTH, "Time difference")
+REGION_ID = SnowflakeBitSegment(settings.SNOWFLAKE_REGION_ID_LENGTH, "Region ID")
+REGION_SEQUENCE = SnowflakeBitSegment(settings.SNOWFLAKE_REGION_SEQUENCE_LENGTH, "Region sequence")
```

Comment:
A purely stylistic note about capturing the order of these four values. A normal way to do it would be this:

```python
VERSION_ID = SnowflakeBitSegment(5, "Version ID")
TIME_DIFFERENCE = SnowflakeBitSegment(32, "Time difference")
REGION_ID = SnowflakeBitSegment(12, "Region ID")
REGION_SEQUENCE = SnowflakeBitSegment(4, "Region sequence")
BIT_SEGMENT_SCHEMA = (VERSION_ID, TIME_DIFFERENCE, REGION_ID, REGION_SEQUENCE)
```

This may be controversial, but I like the following:

```python
BIT_SEGMENT_SCHEMA = (
    VERSION_ID := SnowflakeBitSegment(5, "Version ID"),
    TIME_DIFFERENCE := SnowflakeBitSegment(32, "Time difference"),
    REGION_ID := SnowflakeBitSegment(12, "Region ID"),
    REGION_SEQUENCE := SnowflakeBitSegment(4, "Region sequence"),
)
```

(Btw, `:=` is called the "walrus operator" and it's a little-used language feature. Often used awkwardly for a C-style assignment in the conditional part of a while-loop, but I think the above is actually very nice.)